util/src/test/java/io/grpc/util/AdvancedTlsX509KeyManagerTest.java

```diff
@@ -0,0 +1,175 @@
+/*
```

Comment:
Disclaimer: I'm very much in favor of more testing: )

How do you view the roles of this test file compared to `AdvancedTlsTest.java`? Is the goal of this test file to be the unit tests, whereas `AdvancedTlsTest.java` consists of the e2e-style tests? If yes, should we be comprehensive in this file and ensure that all methods in `AdvancedTlsX509KeyManager` are adequately unit-tested?

Reply:
Yes, `AdvancedTlsTest.java` is testing the approach from e2e perspective, involves both Key and Trust manager, and, IMHO is more useful for the users. 
While looking at it I realized that some of the minor cases (like NPEs) are not covered, and LOC like 
https://github.com/grpc/grpc-java/blob/2bc4306940e7a2dea529be8a6ce536cab0638dc0/netty/src/test/java/io/grpc/netty/AdvancedTlsTest.java#L341

are confusing - no matter what timeout you specify, the initial call for loading creds is synchronous.
So I ended up with some extra unit test coverage of API itself (plus updating javadoc), plus checking that our minimum period is really enforced.
Unit testing all methods and cases will be a)very complicated because of bunch of private/final things (even mockito struggles a lot with it, for example, look at how to emulate a FS reading error of a keyFile) b)not fitting the overall style of the repo (I couldn't see such super detailed tests)

Reply:
> no matter what timeout you specify, the initial call for loading creds is synchronous.

Oh, we should use `FakeClock` for that. Then we can advance time manually within the test and only have the one test thread.

Reply:
> Oh, we should use `FakeClock` for that.

Agreed. Right now there is no test covering time based reloading from file - the current one just relies on the fact that the `updateIdentityCredentialsFromFile(File keyFile, File certFile, long period, TimeUnit unit, ScheduledExecutorService executor)` calls `readAndUpdate` before scheduling the periodic updates - 
https://github.com/grpc/grpc-java/blob/8aaace12eb8697a2dd841d2261b0051a1414f0c8/util/src/main/java/io/grpc/util/AdvancedTlsX509KeyManager.java#L129

I can cover it in a separate PR and do it in probably e2e AdvancedTlsTest.

Reply:
I'm fine with you adding the test in a separate PR, but let's swap to FakeClock now, to never use additional threads. (`new FakeClock().getScheduledExecutorService()`)